tinygraph/tinygraph-rng.c

```diff
@@ -0,0 +1,130 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "tinygraph-hash.h"
+
+/*
+ * Non-cryptographic pseudo-random number generator
+ * based on the PCG design guidelines: a linear
+ * congruential generator with output permutation.
+ *
+ * We adapt the output permutation based on recent
+ * advances from the integer hashing domain.
+ *
+ * See
+ *
+ * - https://www.pcg-random.org
+ *
+ * - https://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf
+ *   Section 6.3.4 "PCG-RXS-M-XS"
+ *
+ * - https://nullprogram.com/blog/2018/07/31/
+ *
+ * - https://github.com/skeeto/hash-prospector/issues/19
+ */
+
+
+// https://nullprogram.com/blog/2018/07/31/
+// https://github.com/skeeto/hash-prospector/issues/19
+static inline uint32_t bitmix(uint32_t v) {
+  v ^= v >> 16;
+  v *= UINT32_C(0x21f0aaad);
+  v ^= v >> 15;
+  v *= UINT32_C(0x735a2d97);
+  v ^= v >> 15;
+
+  return v;
+}
+
+
+typedef struct tinygraph_rng {
+  uint32_t state;
+} tinygraph_rng;
+
+
+tinygraph_rng* tinygraph_rng_construct(void) {
+  tinygraph_rng *out = malloc(sizeof(tinygraph_rng));
+
+  if (!out) {
+    return NULL;
+  }
+
+  // Note: we hard code an initial seed and the increment below
+  // when we step the lcg. Ideally we would take two additional
+  // inputs from the user: a seed (to be used here) and for the
+  // stepping an increment (must be odd) we can then use.
+  //
+  // For the seed the pcg initializes as follows
+  //
+  // state = 0;
+  // step()
+  // state += seed;
+  // step()
+
+  *out = (tinygraph_rng) {
+    .state = UINT32_C(0x46b56677),
+  };
+
+  return out;
+}
+
+
+tinygraph_rng* tinygraph_rng_copy(const tinygraph_rng * const rng) {
+  TINYGRAPH_ASSERT(rng);
+
+  tinygraph_rng *copy = tinygraph_rng_construct();
+
+  if (!copy) {
+    return NULL;
+  }
+
+  copy->state = rng->state;
+
+  return copy;
+}
+
+
+void tinygraph_rng_destruct(tinygraph_rng * const rng) {
+  if (!rng) {
+    return;
+  }
+
+  free(rng);
+}
+
+
+uint32_t tinygraph_rng_random(tinygraph_rng * const rng) {
+  TINYGRAPH_ASSERT(rng);
+
+  // we permute the previous state and step the rng state
+  const uint32_t state = rng->state;
+
+  // use the pcg default multiplier and increment for 32 bit
+  rng->state = rng->state * UINT32_C(747796405) + UINT32_C(2891336453);
+
+  // permute the previous state, and not the updated rng->state
+  return bitmix(state);
```

Comment:
linear congruential generator plus output permutation, where we use the high quality bit mixer